### ISSUE TITLE
parsing failed

### ISSUE THREAD

**Issue by RBerkheimer** [hansec/fortran-language-server#11]
Hi,

Thanks for the module. I have a large f95 project - everything seems to parse fine except for one file (which works in the project). I'd appreciate any ideas - I've tried paring it down, removing intrinsic functions, removing comments, shortening var names, and renaming the file to a shorter name, but it still bombs. 

> Parsing failed for file "/Users/rab25/projects/noaa/asos/src/fortran/main/common/utils/StringConversionUtils.f95"
> Fortran (fortls) /Users/rab25/projects/noaa/asos/

Here's the input/output for debug (seems to run through debug fine):

>Ryans-MacBook-Pro:~ rab25$ python3 /Users/rab25/.atom/packages/autocomplete-fortran/python/parse_fortran.py -d --files=/Users/rab25/projects/noaa/asos/src/fortran/main/common/utils/StringConversionUtils.f95

```
module StringConversionUtils !!! MODULE statement(23)
Found visiblity statement, /Users/rab25/projects/noaa/asos/src/fortran/main/common/utils/StringConversionUtils.f95:28, private :: string_conversion_real, string_conversion_int, string_conversion_complex
Found visiblity statement, /Users/rab25/projects/noaa/asos/src/fortran/main/common/utils/StringConversionUtils.f95:29, private :: string_conversion_logical, string_conversion_int_1d, string_conversion_int_2d
interface string_conversion !!! INTERFACE statement(39)
module procedure string_conversion_real !!! INTERFACE-PRO statement(40)
module procedure string_conversion_int !!! INTERFACE-PRO statement(41)
module procedure string_conversion_complex !!! INTERFACE-PRO statement(42)
module procedure string_conversion_logical !!! INTERFACE-PRO statement(43)
module procedure string_conversion_int_1d !!! INTERFACE-PRO statement(44)
module procedure string_conversion_int_2d !!! INTERFACE-PRO statement(45)
end interface string_conversion !!! END scope(46)
function string_conversion_real(val) result (str) !!! FUNCTION statement(54)
real :: val !!! VARIABLE statement(55)
character(len=200) :: str !!! VARIABLE statement(56)
end function string_conversion_real !!! END scope(61)
function string_conversion_int(val) result (str) !!! FUNCTION statement(64)
integer :: val !!! VARIABLE statement(65)
character(len=200) :: str !!! VARIABLE statement(66)
end function string_conversion_int !!! END scope(71)
function string_conversion_complex(val) result (str) !!! FUNCTION statement(74)
complex :: val !!! VARIABLE statement(75)
character(len=200) :: str !!! VARIABLE statement(76)
end function string_conversion_complex !!! END scope(81)
function string_conversion_logical(val) result (str) !!! FUNCTION statement(84)
logical :: val !!! VARIABLE statement(85)
character(len=5) :: str !!! VARIABLE statement(86)
end function string_conversion_logical !!! END scope(94)
function string_conversion_int_1d(val) result (str) !!! FUNCTION statement(97)
integer, dimension(:) :: val !!! VARIABLE statement(98)
character(len=200) :: str !!! VARIABLE statement(99)
end function string_conversion_int_1d !!! END scope(104)
function string_conversion_int_2d(val) result (str) !!! FUNCTION statement(107)
integer, dimension(:,:)  :: val !!! VARIABLE statement(108)
character(len=200) :: str !!! VARIABLE statement(109)
end function string_conversion_int_2d !!! END scope(114)
end module StringConversionUtils !!! END scope(116)
{"objs":{"stringconversionutils":{"name":"StringConversionUtils","type":1,"desc":"MODULE","fbound":[23,116],"mem":["string_conversion","string_conversion_real","string_conversion_int","string_conversion_complex","string_conversion_logical","string_conversion_int_1d","string_conversion_int_2d"]},"stringconversionutils::string_conversion_real":{"name":"string_conversion_real","type":3,"desc":"CHARACTER(len=200)","fbound":[54,61],"mem":["val","str"],"args":"val"},"stringconversionutils::string_conversion_int":{"name":"string_conversion_int","type":3,"desc":"CHARACTER(len=200)","fbound":[64,71],"mem":["val","str"],"args":"val"},"stringconversionutils::string_conversion_complex":{"name":"string_conversion_complex","type":3,"desc":"CHARACTER(len=200)","fbound":[74,81],"mem":["val","str"],"args":"val"},"stringconversionutils::string_conversion_logical":{"name":"string_conversion_logical","type":3,"desc":"CHARACTER(len=5)","fbound":[84,94],"mem":["val","str"],"args":"val"},"stringconversionutils::string_conversion_int_1d":{"name":"string_conversion_int_1d","type":3,"desc":"CHARACTER(len=200)","fbound":[97,104],"mem":["val","str"],"args":"val"},"stringconversionutils::string_conversion_int_2d":{"name":"string_conversion_int_2d","type":3,"desc":"CHARACTER(len=200)","fbound":[107,114],"mem":["val","str"],"args":"val"},"stringconversionutils::string_conversion":{"name":"string_conversion","type":7,"fbound":[39,46],"mem":["string_conversion_real","string_conversion_int","string_conversion_complex","string_conversion_logical","string_conversion_int_1d","string_conversion_int_2d"]},"stringconversionutils::string_conversion_real::val":{"name":"val","type":6,"fdef":55,"desc":"REAL"},"stringconversionutils::string_conversion_real::str":{"name":"str","type":6,"fdef":56,"desc":"CHARACTER(len=200)"},"stringconversionutils::string_conversion_int::val":{"name":"val","type":6,"fdef":65,"desc":"INTEGER"},"stringconversionutils::string_conversion_int::str":{"name":"str","type":6,"fdef":66,"desc":"CHARACTER(len=200)"},"stringconversionutils::string_conversion_complex::val":{"name":"val","type":6,"fdef":75,"desc":"COMPLEX"},"stringconversionutils::string_conversion_complex::str":{"name":"str","type":6,"fdef":76,"desc":"CHARACTER(len=200)"},"stringconversionutils::string_conversion_logical::val":{"name":"val","type":6,"fdef":85,"desc":"LOGICAL"},"stringconversionutils::string_conversion_logical::str":{"name":"str","type":6,"fdef":86,"desc":"CHARACTER(len=5)"},"stringconversionutils::string_conversion_int_1d::val":{"name":"val","type":6,"fdef":98,"desc":"INTEGER","mods":[21]},"stringconversionutils::string_conversion_int_1d::str":{"name":"str","type":6,"fdef":99,"desc":"CHARACTER(len=200)"},"stringconversionutils::string_conversion_int_2d::val":{"name":"val","type":6,"fdef":108,"desc":"INTEGER","mods":[22]},"stringconversionutils::string_conversion_int_2d::str":{"name":"str","type":6,"fdef":109,"desc":"CHARACTER(len=200)"}},"scopes":["stringconversionutils","stringconversionutils::string_conversion_real","stringconversionutils::string_conversion_int","stringconversionutils::string_conversion_complex","stringconversionutils::string_conversion_logical","stringconversionutils::string_conversion_int_1d","stringconversionutils::string_conversion_int_2d"]}
```

Here's the code:
```

!> @brief Provides functionality for writing log messages of different importance
!! levels to a specified log file and standard out.
!!
!! @details Allowed levels are FATAL, ERROR, WARN, INFO, DEBUG, or TRACE. All log messages
!! will be printed with a timestamp (local time), log level indicator, and the
!! specified message. Log levels allow a developer to limit the messages printed
!! to the log without having to comment them out in the code. For instance, DEBUG
!! messages can be included in the code but only enabled when troubleshooting is
!! required.
!!
!! @copyright
!! THIS SOFTWARE AND ITS DOCUMENTATION ARE CONSIDERED TO BE IN THE PUBLIC
!! DOMAIN AND THUS ARE AVAILABLE FOR UNRESTRICTED PUBLIC USE. THEY ARE
!! FURNISHED "AS IS." THE AUTHORS, THE UNITED STATES GOVERNMENT, ITS
!! INSTRUMENTALITIES, OFFICERS, EMPLOYEES, AND AGENTS MAKE NO WARRANTY,
!! EXPRESS OR IMPLIED, AS TO THE USEFULNESS OF THE SOFTWARE AND
!! DOCUMENTATION FOR ANY PURPOSE. THEY ASSUME NO RESPONSIBILITY (1) FOR
!! THE USE OF THE SOFTWARE AND DOCUMENTATION; OR (2) TO PROVIDE TECHNICAL
!! SUPPORT TO USERS.
!!
!! @author Diana Kantor
!! Modified for use in TaskAPI by Ryan Berkheimer
module StringConversionUtils

implicit none

! Setting helper methods to private
private :: string_conversion_real, string_conversion_int, string_conversion_complex
private :: string_conversion_logical, string_conversion_int_1d, string_conversion_int_2d

!> Converts a non-string to a string, so that a user can
!! easily append a non-string into a log message without
!! having to use complicated string formatting.
!! <br/>
!! ex. `call log_info("This number, "//log_string(my_real)//", will print to the log."))`
!!
!! @param val The value to be converted to a string.
!! @return A string representation of the val passed in.
interface string_conversion
  module procedure string_conversion_real
  module procedure string_conversion_int
  module procedure string_conversion_complex
  module procedure string_conversion_logical
  module procedure string_conversion_int_1d
  module procedure string_conversion_int_2d
end interface string_conversion


contains


!> Converts a real to a string
!! @todo preserve the number of decimal places
function string_conversion_real(val) result (str)
  real :: val
  character(len=200) :: str

  write(str,*) val
  str = trim(adjustl(str))
  return
end function string_conversion_real

!> Converts an integer to a string
function string_conversion_int(val) result (str)
  integer :: val
  character(len=200) :: str

  write(str,*) val
  str = trim(adjustl(str))
  return
end function string_conversion_int

!> Converts a complex number to a string
function string_conversion_complex(val) result (str)
  complex :: val
  character(len=200) :: str

  write(str,*) val
  str = trim(adjustl(str))
  return
end function string_conversion_complex

!> Converts a logical to a string
function string_conversion_logical(val) result (str)
  logical :: val
  character(len=5) :: str

  if(val) then
    str = 'True'
  else
    str = 'False'
  endif
  return
end function string_conversion_logical

!> Converts a 1-D integer array to a string
function string_conversion_int_1d(val) result (str)
  integer, dimension(:) :: val
  character(len=200) :: str

  write(str,*) val
  str = trim(adjustl(str))
  return
end function string_conversion_int_1d

!> Converts a 2-D integer array to a string
function string_conversion_int_2d(val) result (str)
  integer, dimension(:,:)  :: val
  character(len=200) :: str

  write(str,*) val
  str = trim(adjustl(str))
  return
end function string_conversion_int_2d

end module StringConversionUtils

```

**Comment by hansec** [comment]
The parsing error was due to incorrect handling of uppercase characters in object names. I have pushed a fix to master and will include it in the next release, hopefully in the next couple days.

Also, you accidentally tested the parser for the old [autocomplete-fortran Atom package](https://github.com/hansec/autocomplete-fortran). To test the language server parser follow the [instructions here](https://github.com/hansec/fortran-language-server/wiki/Debugging-parser-errors).
